packages/terra-wallet-provider/lib/TerraWalletProvider.ts

```diff
@@ -0,0 +1,193 @@
+import { WalletProvider } from '@liquality/wallet-provider'
+import { Address, BigNumber, Transaction, terra } from '@liquality/types'
+import { addressToString } from '@liquality/utils'
+import { TerraNetwork } from '@liquality/terra-networks'
+import {
+  BlockTxBroadcastResult,
+  Coins,
+  LCDClient,
+  MnemonicKey,
+  Msg,
+  MsgSend,
+  StdTx,
+  StdFee,
+  Wallet
+} from '@terra-money/terra.js'
+
+interface TerraWalletProviderOptions {
+  network: TerraNetwork
+  mnemonic: string
+  derivationPath: string
+}
+
+export default class TerraWalletProvider extends WalletProvider {
+  _network: TerraNetwork
+  _derivationPath: string
+  _addressCache: { [key: string]: Address }
+  private _mnemonic: string
+  private _signer: MnemonicKey
+  private _lcdClient: LCDClient
+  private _wallet: Wallet
+  _accAddressKey: string
+
+  constructor(options: TerraWalletProviderOptions) {
+    const { network, mnemonic, derivationPath } = options
+    super({ network })
+    this._network = network
+    this._mnemonic = mnemonic
+    this._derivationPath = derivationPath
+    this._addressCache = {}
+
+    this._lcdClient = new LCDClient({
+      URL: network.nodeUrl,
+      chainID: network.chainID
+    })
+
+    this._setSigner()
+    this._createWallet(this._signer)
+  }
+
+  async isWalletAvailable(): Promise<boolean> {
+    const addresses = await this.getAddresses()
+    return addresses.length > 0
+  }
+
+  async getAddresses(): Promise<Address[]> {
+    if (this._addressCache[this._mnemonic]) {
+      return [this._addressCache[this._mnemonic]]
+    }
+
+    const wallet = new MnemonicKey({
+      mnemonic: this._mnemonic
+    })
+
+    const result = new Address({
+      address: wallet.accAddress,
+      derivationPath: this._derivationPath,
+      publicKey: wallet.accPubKey
+    })
+
+    this._addressCache[this._mnemonic] = result
+    return [result]
+  }
+
+  async getUsedAddresses(): Promise<Address[]> {
+    return await this.getAddresses()
+  }
+
+  async getUnusedAddress(): Promise<Address> {
+    const addresses = await this.getAddresses()
+    return addresses[0]
+  }
+
+  async signMessage(message: string): Promise<string> {
+    const signed = await this._signer.sign(Buffer.from(message))
+
+    return signed.toString('hex')
+  }
+
+  async getConnectedNetwork(): Promise<TerraNetwork> {
+    return this._network
+  }
+
+  async sendTransaction(sendOptions: terra.TerraSendOptions): Promise<Transaction<terra.InputTransaction>> {
+    const { to, value, messages } = sendOptions
+
+    const data = sendOptions.data as any
+
+    const msgs = []
+
+    if (data) {
+      return this._handleInjection(data)
```

Comment:
Usually we shouldn't refer to some functionality like injection here

What i would do is:
1. Do sendOptions.data as `CreateAndSignTxInput` (whatever the input type to `createAndSignTx is)
2. In this same function spread the sendOptions.data into `createAndSignTx`
3. Only if `to` and `value` are good values, add the `_sendMessage`

I think this way, dont need another function and also `gasPrices` will be defaulted if not exist